### ISSUE TITLE
Demystify input as a folder vs input as a file

### ISSUE THREAD

**Issue by WillBeebe** [bacalhau-project/bacalhau#2413]
Running through David's images of cheese [example](https://youtu.be/gAHaMsTknZM?t=455) I hit a snag. Here is what I was doing.

- upload an image to IPFS
- use the CID of that image as an input
- `bacalhau docker run --input ipfs://QmYjsXqVaN6FzdGjCis6E6iqmhwME93CobCCFnEKXXJTZn ubuntu ${CMD}`
- My script wasn't working, I didn't find anything "inside" the `inputs` folder. 
- I read the [Onboarding Your Docker Workloads](https://docs.bacalhau.org/getting-started/docker-workload-onboarding) docs
- After more scripting I noticed that the "folder" `inputs` was not the right size... it must be a file 🤯 
- Once I treated `inputs` as a file and not a folder, my scripts worked.

This was confusing to me and I spent 80% of my time working through this issue and not using Bacalhau. If there was an example that helped clarify that this can be a file _or_ a folder that would be helpful. Cheers!

**Comment by simonwo** [comment]
Agreed, multiple people are caught out by this. We should decide on a single policy that is always true (e.g. single file input ends up in a folder) and implement that. Thanks for the push!

**Comment by simonwo** [comment]
Another example from https://bacalhauproject.slack.com/archives/C052BAZJVEU/p1689087168410349?thread_ts=1688986814.683239&cid=C052BAZJVEU: 
```
./bacalhau docker run \
 --wait \
 --wait-timeout-secs 100 \
 -i ipfs://qmnd9istxcsekdre7escphwthk3hw55xt1wg3wehefnp3t/:/input_images \
 dpokidov/imagemagick:7.1.0-47-ubuntu \
-- bash -c "magick mogrify -resize 100x100 -quality 100 -path /outputs /input_images/*.jpeg"
```
Again, the CID was to a single file but the user was expecting this to be a folder.

**Comment by simonwo** [comment]
More context: #2364.
> I'm currently looking at StorageProviders (2 :yak: deep) and it looks like in the S3 provider the PrepareStorage returns a Volume pointing to a target directory - this makes sense.  However, in the urldownload provider (used by --input-urls) PrepareStorage returns a Volume pointing to a single file.  This means that when I download https://somehost.somewhere.com/data.txt it appears as /inputs instead of /inputs/data.txt.

**Comment by simonwo** [comment]
More context: #857. Different input engines treat things differently.

**Comment by simonwo** [comment]
> https://bacalhauproject.slack.com/archives/C052BAZJVEU/p1666983588565149
> Hey! Is there a way to mount just a single file from an IPFS directory into a Bacalhau job? i.e. QmTjc3EhaN1LP3sR8wEm2kiNvBaaXjUyKhZNNncED6dPAn/3224.json instead of mounting the directory with all of the files?

So users _do_ want to be able to mount single files in some cases.

**Comment by simonwo** [comment]
More fun: #1374. URL downloader is not clear about the name of the final file.

**Comment by simonwo** [comment]
So what is the user need here? The need is that:

* As a user, I have an expectation/mental model of how my input data will be exposed to the executor and I write code for the executor that matches that expectation. 
* So I need that expectation to be met, i.e. if I'm expecting a folder, I want to receive a folder in the executor.

And:

* When that expectation is not met, the error messages that I receive need to make it clear what the problem is (i.e. not that the data input "didn't work", but just the expectation about how it's presented was not met.

So what can we do about this?

## 1. Always present a folder
Nice and simple rule, even single files get wrapped into a folder. We teach users to always expect a folder.

Except sometimes this is not what the user expects, clearly if I want to mount `src=file:///var/run/osquery.sock,dst=/osquery.sock` I am not expecting to find a structure  `/osquery.sock/osquery.sock` or equivalent. 

Similarly the naming issues with URL downloads exist because we always treat the input path as a folder and thus have to synthesise a name for the final file.

### 1a. Only do this when the destination mount point is implicit
This might work however if we limit this behaviour to the "simple" `--input` flag where we don't manually specify `src=...,dst=...`. In that world the dst is always mapped to `/inputs` and so we land the idea that in the implicit case `/inputs` is always a directory. So doing:

* `$ bacalhau ... run --input file:///var/run/osquery.sock` would result in `/inputs/osquery.sock` (implicit)
* `$ bacalhau ... run --input file:///var/run/osquery.sock:/cool_beans` would result in `/cool_beans/stuff.txt` (implicit-ish)
* `$ bacalhau ... run --input src=file:///var/run/osquery.sock,dst=/osquery.sock` would result in `/osquery.sock` (explicit)

This does mean that `-input ...` is not equivalent to `--input src=...,dst=/inputs` – is this wise?

### 1b. Add a flag for explicitly controlling directory wrapping

So that `--input` can still be equivalent to an extended flag, e.g. these are equivalent:

* `$ bacalhau ... run --input file:///var/run/osquery.sock`
* `$ bacalhau ... run --input src=file:///var/run/osquery.sock,dst=/inputs,wrap=auto` 

And would both result in `/inputs/osquery.sock`. The options would be something like:

* `auto` – if the input is a file, wrap it in a directory
* `off` – if the input it a file, mount it directly as a file

### 1c. Only do this for IPFS

It seems like most of the confusion seems to come from mounting CIDs and expecting them to be folders, so we could apply some of the above to only specific storage sources.

## 2. Try and guess the user's intent.

 E.g. if they are mounting a file at a path that "looks like" a file, then mount the file, else wrap it in a directory. E.g. 

* `$ bacalhau ... run --input file:///var/run/osquery.sock` would result in `/inputs/osquery.sock`
* `$ bacalhau ... run --input src=file:///var/run/osquery.sock,dst=/inputs` would result in `/inputs/osquery.sock`
* `$ bacalhau ... run --input src=file:///var/run/osquery.sock,dst=/inputs.sock` would result in `/inputs.sock`

Feels a bit magic and prone to issues as there is not a widely recognised way of doing this, e.g. `/etc/conf.d` looks like a file but `.d` is normally a directory.

## 3. Add some examples early in the learning pathway that use single-file input

A non-technical solution – just make sure our users have the right expectations and are able to understand that CIDs come in both file and directory flavours.

Not super clear if our users are reading the examples though! And plenty of evidence in the above links of directly explaining "this is a file, that is a folder" and not getting very far. So this is maybe a hard content problem.

**Comment by simonwo** [comment]
[Walid's comment](https://bacalhauproject.slack.com/archives/C052BAZJVEU/p1689197250054459?thread_ts=1689176939.514469&cid=C052BAZJVEU):

> If am I understood your proposal correctly, and assuming an input `-i ipfs://cid/:/my_dest`
> 
> -   If the input was detected as folder, then it will be mounted at /my_dest
> -  If the input was detected as file, then it will be mounted at /my_dest/cid
> -   If the input was was a file, but wrapping is off, it will be mounted at /my_dest
> 
> I am just more in favour of explicit and more predictable behaviour, but I am okay if you believe there is value in doing what you are proposing. And yes, in my suggestion we should fail if the user says `file` mode but the input wasn't a file 
>
> also if  `wrap=auto` by default in this format `--input ipfs://..`. , then it should also be auto in `--input src=ipfs://` format and when submitting yaml jobs through `bacalhau create`

**Comment by phillipleblanc** [comment]
My 2 cents as a Bacalhau user.

The majority of the time for my use-case, I want to mount a single file (or specify multiple single files to mount at different explicit paths). Even when I need to work with multiple files, I will archive them first (i.e. zip or tar) and upload that to IPFS. I do have a use-case to mount a local directory into a directory in a bacalhau job, but I don't currently ever mount IPFS directories directly into Bacalhau. Mainly because outside of IPFS, it is much harder to work with directories (unless you are working on a local filesystem). So we tend to archive files together and pass the archive around, and when we need to run Bacalhau jobs we will run it directly on the archive file.

**Comment by simonwo** [comment]
Thanks @phillipleblanc! I think you have prompted that we need to just look at what inputs people are running over – yours may be the common case.

Would using the `src=…,dst=...` format work for you?

**Comment by phillipleblanc** [comment]
@simonwo We aren't using the CLI to submit jobs, we use the API from Golang and specify the inputs in the job spec. But `src=...,dst=...` makes sense to me. Maybe if people leave it off then the CLI can emit a message that shows how it translated it into the src,dst format to make it explicit?

**Comment by rossjones** [comment]
More context: Recently we found issues where a user was mounting both http and ipfs and  unless they were both in subdirectories of /inputs then we would get errors about docker failing to mount volumes.  

The solution was for `http->/inputs/notebook/`, and ipfs to`/inputs/data/` but this wasn't obvious to the user.

**Comment by Nasfame** [comment]
My thoughts are that the behaviour should be like the `cp` command

Imaging downloading to the temp directory and cp from temp the directory to the specified dst
